lib/grandpa/catch-up.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+const catchUpResponseTimeout = 5 * time.Second
+
+type catchUp struct {
+	lock sync.Mutex // applied on requestsSent
+
+	requestsSent map[peer.ID]CatchUpRequest
+	bestResponse *atomic.Value // *CatchUpResponse
+
+	catchUpResponseCh chan *CatchUpResponse
+	waitingOnResponse *atomic.Value
+	grandpa           *Service
+}
+
+func (c *catchUp) do(to peer.ID, round uint64, setID uint64) error {
+	defer c.waitingOnResponse.Store(false)
+
+	if err := c.sendCatchUpRequest(
+		to, newCatchUpRequest(round, setID),
+	); err != nil {
+		return fmt.Errorf("failed to send catch up request: %w", err)
+	}
+
+	logger.Debugf("successfully sent a catch up request to node %s, for round number %d and set ID %d",
+		to, round, setID)
+
+	c.waitingOnResponse.Store(true)
```

Comment:
you call this `c.waitingOnResponse.Store(true)` function in `sendCatchUpRequest` and call here as well, should you remove this one?